### PR TITLE
Add grpc_retry_test

### PR DESCRIPTION
**Pull request by ionagamed**
Signed-off-by: Leonid Lygin <ionagamed@gmail.com>

### Description of the Change

This adds a test for gRPC retries. 

### Benefits

Now this functionality is tested

### Possible Drawbacks 

These two tests require 70s to run on my machine. Maybe it is somehow possible to override the timeouts, so that the functionality can be tested with fewer time

### JIRA issue

https://jira.hyperledger.org/browse/IR-586

### CODE REVIEW

**Comment by MBoldyrev** [comment]
> These two tests require 70s to run on my machine.

70s

File: test/integration/grpc_retry_test.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+ private:
+  int attempts_ = 0;
+  int max_attempts_;
+};
+
+namespace {
+  constexpr int kAttemptsForSuccess = 3;  // 5 max attempts at reaching
+  constexpr int kAttemptsForFailure = 5;  // 5 max attempts at reaching
```
Comment (MBoldyrev) [inline test/integration/grpc_retry_test.cpp]: `kAttemptsForSuccess` and `kAttemptsForFailure` should differ by 1 to test edge cases

File: test/integration/grpc_retry_test.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    iroha::protocol::Query query;
+    iroha::protocol::QueryResponse response;
+
+    grpc::ClientContext client_context;
+
+    auto status = client->Find(&client_context, query, &response);
+
+    SCOPED_TRACE(status.error_code());
+    SCOPED_TRACE(status.error_message());
+    SCOPED_TRACE(status.error_details());
```
Comment (MBoldyrev) [inline test/integration/grpc_retry_test.cpp]: Why `SCOPED_TRACE`s? I see no subroutines. You can print a nice message with `<<` to ASSERT_EQ.

File: test/integration/grpc_retry_test.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      : iroha::torii::QueryService(
+          nullptr, nullptr, nullptr, getTestLogger("MockQueryService")),
+        max_attempts_(max_attempts) {}
+
+  grpc::Status Find(grpc::ServerContext *context,
+                    const iroha::protocol::Query *request,
+                    iroha::protocol::QueryResponse *response) override {
+    if (attempts_ < max_attempts_) {
+      attempts_++;
+      return grpc::Status(grpc::StatusCode::UNAVAILABLE, "");
```
Comment (MBoldyrev) [inline test/integration/grpc_retry_test.cpp]: we could also set a message here to some constant and check it on the client's side

File: test/integration/grpc_retry_test.cpp (hunk 1 above)
Comment (ionagamed) [reply test/integration/grpc_retry_test.cpp]: Fixed in the latest version

File: test/integration/grpc_retry_test.cpp (hunk 2 above)
Comment (ionagamed) [reply test/integration/grpc_retry_test.cpp]: Removed these in the latest version, they were not particularly useful anyway

File: test/integration/grpc_retry_test.cpp (hunk 3 above)
Comment (ionagamed) [reply test/integration/grpc_retry_test.cpp]: Done in the latest version

**Comment by ionagamed** [comment]
Ready for a re-review, removed the `SCOPED_TRACE`s, added a custom error message, and re-formatted parts where clang-format CI job wasn't happy

File: irohad/torii/query_service.hpp [hunk 4]
```diff
@@ -60,9 +60,10 @@ namespace iroha {
       void Find(iroha::protocol::Query const &request,
                 iroha::protocol::QueryResponse &response);
 
-      grpc::Status Find(grpc::ServerContext *context,
-                        const iroha::protocol::Query *request,
-                        iroha::protocol::QueryResponse *response) override;
+      virtual grpc::Status Find(
```
Comment (lebdron) [inline irohad/torii/query_service.hpp]: `virtual` should not be required because the method is already marked as `override`.

File: test/integration/CMakeLists.txt [hunk 5]
```diff
@@ -9,3 +9,10 @@ add_subdirectory(validation)
 if (USE_BTF)
   add_subdirectory(binary)
 endif()
+
+
+addtest(grpc_retry_test grpc_retry_test.cpp)
+target_link_libraries(grpc_retry_test
+    application
```
Comment (lebdron) [inline test/integration/CMakeLists.txt]: `application` should not be required here, because only several components are used in the test.

File: test/integration/grpc_retry_test.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+#include "framework/test_logger.hpp"
+#include "main/server_runner.hpp"
+#include "network/impl/grpc_channel_builder.hpp"
+#include "qry_responses.pb.h"
+#include "queries.pb.h"
+#include "torii/query_service.hpp"
+
+namespace {
+  constexpr auto kErrorMessage = "this is a test error message";
+};
```
Comment (lebdron) [inline test/integration/grpc_retry_test.cpp]:
```suggestion
}
```

File: test/integration/grpc_retry_test.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+class MockQueryService : public iroha::torii::QueryService {
```
Comment (lebdron) [inline test/integration/grpc_retry_test.cpp]:
```suggestion
class MockQueryService : public iroha::protocol::QueryService_v1::Service {
```

File: test/integration/grpc_retry_test.cpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ private:
+  int attempts_ = 0;
+  int max_attempts_;
+};
+
+namespace {
+  constexpr int kAttemptsForSuccess = 4;  // 5 max attempts at reaching
+  constexpr int kAttemptsForFailure = 5;  // 5 max attempts at reaching
+  constexpr auto kListenIP = "127.0.0.1";
+  constexpr auto kListenPort = 31337;
```
Comment (lebdron) [inline test/integration/grpc_retry_test.cpp]: It should be possible to use `0` so that it will assign any port that is not used.

File: test/integration/grpc_retry_test.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    auto status = client->Find(&client_context, query, &response);
+
+    ASSERT_EQ(status.error_code(), code);
+    ASSERT_EQ(status.error_message(), message);
+  }
+
+}  // namespace
+
+TEST(GrpcRetryTest, GrpcRetrySuccessTest) {
```
Comment (lebdron) [inline test/integration/grpc_retry_test.cpp]: Please add given-when-then to both tests and mention how constants are used.

File: test/integration/grpc_retry_test.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#include <gtest/gtest.h>
+
+#include <grpcpp/server_context.h>
+#include <grpcpp/support/status.h>
+#include "framework/test_logger.hpp"
+#include "main/server_runner.hpp"
+#include "network/impl/grpc_channel_builder.hpp"
+#include "qry_responses.pb.h"
+#include "queries.pb.h"
+#include "torii/query_service.hpp"
```
Comment (lebdron) [inline test/integration/grpc_retry_test.cpp]: Torii is not used anymore and can be removed.

**Comment by ionagamed** [comment]
@MBoldyrev `UNAVAILABLE` was added to the list in order to fix a weird synchronisation bug when a network only has two nodes (https://jira.hyperledger.org/browse/IR-586)

**Comment by MBoldyrev** [comment]
> @MBoldyrev `UNAVAILABLE` was added to the list in order to fix a weird synchronisation bug when a network only has two nodes (https://jira.hyperledger.org/browse/IR-586)

It is not directly related to this PR IIUYR. And it appears not a trivial change, because simply adding it to the list of retryable statuses increases the time taken by integration tests a lot. So I'd better add the test in this PR, and addressed the new status retrying in another (which I'm already working on).
